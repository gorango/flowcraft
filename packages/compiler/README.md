# `@flowcraft/compiler`

[![NPM Version](https://img.shields.io/npm/v/@flowcraft/compiler.svg)](https://www.npmjs.com/package/@flowcraft/compiler)
[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT)

The `@flowcraft/compiler` is an optional, opt-in toolchain that transforms intuitive, imperative TypeScript code into a declarative Flowcraft workflow blueprint. It gives you the best of both worlds: a natural developer experience using standard language features and a powerful, serializable, and predictable graph-based execution model at runtime.

### The Problem: Declarative Overhead

The standard way to build a workflow in Flowcraft is with the `createFlow()` fluent API, which is powerful for defining graphs explicitly. However, for complex orchestration with loops, branches, and error handling, this can feel verbose compared to writing standard code.

**Fluent API (Declarative):**
```typescript
const flow = createFlow('my-loop')
  .node('start', getInitialPage)
  .node('controller', isPageAvailable, { config: { joinStrategy: 'any' } })
  .node('fetch', fetchPage)
  .node('process', processItems)
  .edge('start', 'controller')
  .edge('controller', 'fetch', { action: 'continue' })
  .edge('fetch', 'process')
  .edge('process', 'controller'); // Loop back
```

### The Solution: Imperative DX, Declarative Runtime

The compiler lets you write the same logic using a standard `while` loop. The toolchain handles the transformation to the declarative graph behind the scenes.

**Compiler Input (Imperative):**
```typescript
/** @step */
async function getInitialPage() {
  return { page: 1 };
}

/** @step */
async function fetchPage(params: { page: number }) {
  return { items: [/* ... */] };
}

/** @step */
async function processItems(params: { items: any[] }) {
  // Process items...
}

/** @flow */
async function myLoop(context: IAsyncContext) {
  let page = await getInitialPage();
  while (page !== null) {
    const items = await fetchPage({ page });
    await processItems({ items });
    page = await context.get('nextPage'); // State updated by nodes
  }
}
```

## Key Features

-   **Imperative DX**: Write workflows using standard `async/await`, `if/else`, `while`, `try/catch`, and `Promise.all`.
-   **Explicit Step Declaration**: Mark durable operations with `/** @step */` or `/** @flow */` JSDoc tags to ensure only intended functions become part of your workflow graph.
-   **Zero-Syntax**: No framework-specific keywords to learn. Orchestration is defined by native JavaScript control flow, guided by simple JSDoc markers.
-   **Compile-Time Type Safety**: The compiler leverages the TypeScript TypeChecker to validate the data flowing between your nodes, catching type errors before you even run your code.
-   **Composable by Default**: Create subflows simply by importing and `await`-ing another function marked with `/** @flow */`.
-   **Full Tooling Integration**: The generated blueprints include source location metadata, enabling visualization tools to link a graph node directly back to your source code.
-   **Optional & Backwards-Compatible**: Use the compiler for new, complex flows while maintaining existing workflows built with the fluent API.

## Installation

Install the compiler as a development dependency in your project.

```bash
npm install --save-dev @flowcraft/compiler
```

You will also need `flowcraft` and `typescript`:

```bash
npm install flowcraft
npm install --save-dev typescript
```

## Usage

Using the compiler is a simple, four-step process that fits into any standard build system.

### Step 1: Write Your Workflow and Step Functions

Create your step functions as regular exported `async` functions, marking each with the `/** @step */` JSDoc comment. Then, create a main workflow function that orchestrates them. Mark the orchestrator with the `/** @flow */` JSDoc comment.

`./src/steps.ts`
```typescript
// These are our atomic "step" functions
/** @step */
export async function createCart(params: { userId: string }) {
  console.log(`Creating cart for user ${params.userId}...`);
  return { cartId: 'cart-123' };
}

/** @step */
export async function addItems(params: { cartId: string; items: string[] }) {
  console.log(`Adding items to cart ${params.cartId}...`);
  return { success: true };
}

/** @step */
export async function processPayment(params: { cartId: string; token: string }) {
  console.log(`Processing payment for cart ${params.cartId}...`);
  return { transactionId: 'txn-xyz' };
}
```

`./src/checkout.ts`
```typescript
import { createCart, addItems, processPayment } from './steps';

/**
 * This is a "flow function". The compiler will transform it
 * into a WorkflowBlueprint.
 * @flow
 */
export async function checkoutFlow(context: any) {
  const userId = await context.get('userId');
  const items = await context.get('items');
  const token = await context.get('paymentToken');

  const cart = await createCart({ userId });
  await addItems({ cartId: cart.cartId, items });
  const payment = await processPayment({ cartId: cart.cartId, token });

  return payment;
}
```

### Step 2: Create a Build Script

Create a script that invokes the compiler. This can be a simple Node.js file.

`./scripts/compile-flows.js`
```javascript
import { compileProject } from '@flowcraft/compiler';
import path from 'node:path';
import fs from 'node:fs/promises';

async function build() {
  console.log('Compiling Flowcraft workflows...');

  // 1. Define project entry points and tsconfig path
  const entryPoints = [path.resolve('./src/checkout.ts')];
  const tsConfigPath = path.resolve('./tsconfig.json');

  // 2. Compile the project
  const { blueprints, registry, diagnostics, manifestSource } = compileProject(
    entryPoints,
    tsConfigPath,
  );

  // 3. Check for errors
  if (diagnostics.some(d => d.severity === 'error')) {
    console.error('Compilation failed with errors:');
    diagnostics.forEach(d => {
      console.error(`- ${d.file}:${d.line}:${d.column} - ${d.message}`);
    });
    process.exit(1);
  }

  // 4. Write the manifest file
  const manifestPath = path.resolve('./dist/flowcraft.manifest.js');
  await fs.mkdir(path.dirname(manifestPath), { recursive: true });
  await fs.writeFile(manifestPath, manifestSource);

  console.log(`âœ… Compilation successful! Manifest written to ${manifestPath}`);
  console.log(`Discovered ${Object.keys(blueprints).length} blueprints and ${Object.keys(registry).length} step functions.`);
}

build();
```

### Step 3: Run the Build Script

Execute the script from your terminal. You can also add this to the `scripts` section of your `package.json`.

```bash
node ./scripts/compile-flows.js
```

This generates a manifest file containing your blueprints and a registry of your step functions.

`./dist/flowcraft.manifest.js` (Generated File)
```javascript
// Generated by @flowcraft/compiler
import { createCart, addItems, processPayment } from '../src/steps';
// ... other imports

import type { NodeImplementation, WorkflowBlueprint } from 'flowcraft';

export const registry = {
  'createCart': createCart,
  'addItems': addItems,
  'processPayment': processPayment
};

export const blueprints = {
  'checkoutFlow': {
    "id": "checkoutFlow",
    "nodes": [/* ... */],
    "edges": [/* ... */]
  }
};
```

### Step 4: Use the Generated Manifest

Now, you can import the `blueprints` and `registry` from the generated manifest and pass them directly to the `FlowRuntime`.

`./src/main.ts`
```typescript
import { FlowRuntime } from 'flowcraft';
import { blueprints, registry } from '../dist/flowcraft.manifest.js';

async function main() {
  const runtime = new FlowRuntime({
    blueprints, // All compiled blueprints
    registry,   // All discovered step functions
  });

  const checkoutBlueprint = blueprints.checkoutFlow;
  const initialContext = {
    userId: 'user-42',
    items: ['item-a', 'item-b'],
    paymentToken: 'tok_1234',
  };

  const result = await runtime.run(checkoutBlueprint, initialContext);

  console.log('Workflow completed with status:', result.status);
  console.log('Final context:', result.context);
}

main();
```

## Authoring Flows: The Rules

To make the magic possible, you must follow a few simple rules when writing your flow functions.

#### 1. The `/** @flow */` Marker
Any `async` function that you want the compiler to transform into a blueprint **must** have the `/** @flow */` JSDoc tag. This is how the compiler discovers your flows.

#### 2. Step Functions
Any `async` function that is `await`-ed from within a flow function must be explicitly marked as a "step" with the `/** @step */` JSDoc tag. Only functions marked with `/** @step */` or `/** @flow */` are considered durable operations that can be part of a workflow graph. Attempting to await a regular `async` function will result in a compile-time error.

#### 3. Supported JavaScript Syntax

The compiler translates the following standard JavaScript syntax into graph structures:

| Syntax | Generated Flowcraft Pattern |
| :--- | :--- |
| `await step()` | A node and a sequential edge. |
| `if (condition) { ... } else { ... }` | A fork with conditional edges and a merge point with `joinStrategy: 'any'`. |
| `while (condition) { ... }` | A `loop-controller` node with a cyclical graph structure. |
| `for (const item of items) { ... }`| A `loop-controller` node (de-sugared into a `while` loop). |
| `try { ... } catch { ... }` | A fallback path. All nodes in the `try` block are configured to fallback to the first node in the `catch` block. |
| `await Promise.all([ ... ])` | A scatter-gather pattern where parallel branches merge at a successor node with `joinStrategy: 'all'`. |

#### 4. Unsupported Syntax

> **Note:** To ensure predictable graph generation, some imperative features are not supported inside a flow function. Using them will result in a compile-time error.

-   **`break` and `continue`**: Not supported within loops. Loop exit is controlled solely by the loop's condition.
-   **`finally`**: The `finally` block in a `try/catch/finally` statement is not supported.
-   **Complex Assignments**: Variable assignments from `await` calls must be simple `const` or `let` declarations. Re-assigning variables from multiple branches can lead to unpredictable graphs and is disallowed.

## Advanced Concepts

#### Subflows
Creating a subflow is completely natural. Simply define another flow with `/** @flow */`, import it, and `await` it from your main flow. The compiler will automatically generate a `subflow` node.

`./src/main-flow.ts`
```typescript
import { subFlow } from './sub-flow';

/** @step */
export async function doFirstStep() {
  return { result: 'first' };
}

/** @step */
export async function doLastStep() {
  return { result: 'last' };
}

/** @flow */
export async function mainFlow() {
  await doFirstStep();
  await subFlow(); // This becomes a 'subflow' node
  await doLastStep();
}
```

#### Compile-Time Type Safety
The compiler analyzes your code with the full power of the TypeScript engine. If you pass data of the wrong type between steps, the compiler will catch it and issue a diagnostic.

**Example:**

`./src/steps.ts`
```typescript
/** @step */
export async function processData(params: { value: number }) {
  return params.value * 2;
}```

`./src/my-flow.ts`
```typescript
/** @flow */
export async function typeErrorFlow() {
  // Compiler Error: 'value' expects type 'number', but receives 'string'
  await processData({ value: 'this is not a number' });
}
```

**Compiler Output:**
```
- src/my-flow.ts:4:3 - Type error in call to 'processData': argument of type 'string' is not assignable to parameter 'value' of type 'number'
```

## API Reference

#### `compileProject(entryFilePaths, tsConfigPath)`
-   `entryFilePaths: string[]`: An array of absolute paths to your entry-point files (your main flow functions).
-   `tsConfigPath: string`: An absolute path to your project's `tsconfig.json` file.
-   **Returns:** `CompilationOutput`
    -   `blueprints: Record<string, WorkflowBlueprint>`: A map of flow names to their generated blueprints.
    -   `registry: Record<string, { importPath: string, exportName: string }>`: A map of step names to their import locations.
    -   `diagnostics: CompilationDiagnostic[]`: An array of errors or warnings found during compilation.
    -   `manifestSource: string`: The generated source code for the manifest file.
