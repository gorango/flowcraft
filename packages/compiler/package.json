{
	"name": "@flowcraft/compiler",
	"type": "module",
	"version": "1.0.0",
	"description": "A compiler for transforming imperative TypeScript into declarative Flowcraft workflows",
	"license": "MIT",
	"homepage": "https://flowcraft.js.org",
	"repository": {
		"type": "git",
		"url": "https://github.com/gorango/flowcraft.git"
	},
	"bugs": "https://github.com/gorango/flowcraft/issues",
	"keywords": [
		"flowcraft",
		"compiler",
		"workflow",
		"typescript"
	],
	"exports": {
		".": {
			"types": "./dist/index.d.ts",
			"import": "./dist/index.js"
		}
	},
	"files": [
		"dist"
	],
	"scripts": {
		"dev": "tsup --watch",
		"build": "tsup",
		"typecheck": "tsc --noEmit",
		"test": "vitest run",
		"test:watch": "vitest",
		"test:coverage": "vitest run --coverage"
	},
	"dependencies": {
		"flowcraft": "workspace:*",
		"resolve": "^1.22.11",
		"typescript": "^5.9.3"
	},
	"devDependencies": {
		"@types/node": "^24.9.1",
		"@types/resolve": "^1.20.6",
		"@vitest/coverage-v8": "3.2.4",
		"tsup": "^8.5.0",
		"vitest": "^4.0.4"
	}
}
