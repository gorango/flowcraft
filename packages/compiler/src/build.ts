import fs from 'node:fs/promises'
import path from 'node:path'
import { compileProject } from './index'

export interface CompileFlowsOptions {
	entryPoints?: string[]
	tsConfigPath?: string
	manifestPath?: string
}

// This is the generic, reusable function
export async function buildFlows(options: CompileFlowsOptions = {}) {
	const projectRoot = process.cwd()
	const entryPoints = options.entryPoints ?? [path.resolve(projectRoot, 'src/index.ts')]
	const tsConfigPath = options.tsConfigPath ?? path.resolve(projectRoot, 'tsconfig.json')
	const manifestPath = options.manifestPath ?? path.resolve(projectRoot, 'dist/flowcraft.manifest.js')

	console.log('Compiling Flowcraft workflows...')

	try {
		const { diagnostics, manifestSource } = compileProject(entryPoints, tsConfigPath)

		if (diagnostics.some((d) => d.severity === 'error')) {
			console.error('❌ Flowcraft compilation failed:')
			diagnostics
				.filter((d) => d.severity === 'error')
				.forEach((d) => {
					console.error(`  - ${path.relative(projectRoot, d.file)}:${d.line}:${d.column} - ${d.message}`)
				})
			throw new Error('Flowcraft compilation failed.')
		}

		await fs.mkdir(path.dirname(manifestPath), { recursive: true })
		await fs.writeFile(manifestPath, manifestSource)
		console.log(`✅ Flowcraft compilation successful! Manifest: ${path.relative(projectRoot, manifestPath)}`)
	} catch (error) {
		console.error('❌ An unexpected error occurred during Flowcraft compilation:')
		// Don't log the whole error object unless verbose logging is on
		console.error(error instanceof Error ? error.message : String(error))
		// Re-throw to fail the build process
		throw error
	}
}
