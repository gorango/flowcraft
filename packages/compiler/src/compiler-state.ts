import type { EdgeDefinition, NodeDefinition } from 'flowcraft'
import * as ts from 'typescript'
import type { Scope } from './types'

export class CompilerState {
	private cursor: string | null = null
	private nodes: NodeDefinition[] = []
	private edges: EdgeDefinition[] = []
	private scopes: Scope[] = []
	private fallbackScope: string | null = null
	private usageCounts: Map<string, number> = new Map()
	private pendingBranches: { ends: string[]; joinStrategy: 'any' | 'all' } | null = null
	private pendingForkEdges: { source: string; condition: string }[] = []
	private loopScopeStack: { controllerId: string; breakTargetId: string }[] = []

	constructor() {
		// Push initial scope
		this.scopes.push({ variables: new Map() })
	}

	getCursor(): string | null {
		return this.cursor
	}

	setCursor(nodeId: string | null): void {
		this.cursor = nodeId
	}

	getNodes(): NodeDefinition[] {
		return this.nodes
	}

	getEdges(): EdgeDefinition[] {
		return this.edges
	}

	getScopes(): Scope[] {
		return this.scopes
	}

	getFallbackScope(): string | null {
		return this.fallbackScope
	}

	setFallbackScope(scope: string | null): void {
		this.fallbackScope = scope
	}

	getUsageCounts(): Map<string, number> {
		return this.usageCounts
	}

	getPendingBranches(): { ends: string[]; joinStrategy: 'any' | 'all' } | null {
		return this.pendingBranches
	}

	setPendingBranches(branches: { ends: string[]; joinStrategy: 'any' | 'all' } | null): void {
		this.pendingBranches = branches
	}

	getPendingForkEdges(): { source: string; condition: string }[] {
		return this.pendingForkEdges
	}

	pushScope(scope: Scope): void {
		this.scopes.push(scope)
	}

	popScope(): Scope | undefined {
		return this.scopes.pop()
	}

	addNodeAndWire(nodeDef: NodeDefinition, node: ts.Node, sourceFile: ts.SourceFile, typeChecker: ts.TypeChecker): void {
		this.nodes.push(nodeDef)
		if (this.pendingBranches) {
			for (const end of this.pendingBranches.ends) {
				this.edges.push({
					source: end,
					target: nodeDef.id,
					_sourceLocation: this.getSourceLocation(node, sourceFile),
				})
			}
			nodeDef.config = {
				...nodeDef.config,
				joinStrategy: this.pendingBranches.joinStrategy as 'all' | 'any',
			}
			this.pendingBranches = null
		}
		if (this.pendingForkEdges.length > 0) {
			for (const forkEdge of this.pendingForkEdges) {
				this.edges.push({
					source: forkEdge.source,
					target: nodeDef.id,
					condition: forkEdge.condition,
					_sourceLocation: this.getSourceLocation(node, sourceFile),
				})
			}
			this.pendingForkEdges = []
		}
		if (this.cursor) {
			const edge: any = {
				source: this.cursor,
				target: nodeDef.id,
				_sourceLocation: this.getSourceLocation(node, sourceFile),
			}
			// If source is a loop-controller, add break action
			const sourceNode = this.nodes.find((n) => n.id === this.cursor)
			if (sourceNode && sourceNode.uses === 'loop-controller') {
				edge.action = 'break'
			}
			this.edges.push(edge)
		}
		this.cursor = nodeDef.id

		// Map variable to node output if it's a VariableDeclaration
		const parent = node.parent
		if (ts.isVariableDeclaration(parent) && parent.name && ts.isIdentifier(parent.name)) {
			const varName = parent.name.text
			const returnType = typeChecker.getTypeAtLocation(node)
			this.scopes[this.scopes.length - 1].variables.set(varName, {
				nodeId: nodeDef.id,
				type: returnType,
			})
		}
	}

	addNode(nodeDef: NodeDefinition): void {
		this.nodes.push(nodeDef)
	}

	addEdge(edge: EdgeDefinition): void {
		this.edges.push(edge)
	}

	addPendingForkEdge(source: string, condition: string): void {
		this.pendingForkEdges.push({ source, condition })
	}

	pushLoopScope(scope: { controllerId: string; breakTargetId: string }): void {
		this.loopScopeStack.push(scope)
	}

	popLoopScope(): void {
		this.loopScopeStack.pop()
	}

	getCurrentLoopScope(): { controllerId: string; breakTargetId: string } | undefined {
		return this.loopScopeStack[this.loopScopeStack.length - 1]
	}

	incrementUsageCount(name: string): number {
		const count = (this.usageCounts.get(name) || 0) + 1
		this.usageCounts.set(name, count)
		return count
	}

	setUsageCounts(counts: Map<string, number>): void {
		this.usageCounts = new Map(counts)
	}

	private getSourceLocation(node: ts.Node, sourceFile: ts.SourceFile): import('flowcraft').SourceLocation {
		const { line, character } = sourceFile.getLineAndCharacterOfPosition(node.getStart())
		return {
			file: sourceFile.fileName,
			line: line + 1,
			column: character + 1,
		}
	}
}
