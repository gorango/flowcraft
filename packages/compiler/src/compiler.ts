import * as path from 'node:path'
import type { WorkflowBlueprint } from 'flowcraft'
import * as ts from 'typescript'
import { FlowAnalyzer } from './flow-analyzer'
import type { CompilationOutput, FileAnalysis } from './types'

export class Compiler {
	private program: ts.Program
	private typeChecker: ts.TypeChecker
	public fileCache: Map<string, FileAnalysis> = new Map()

	constructor(tsConfigPath: string) {
		const config = ts.readConfigFile(tsConfigPath, ts.sys.readFile)
		const parsed = ts.parseJsonConfigFileContent(config.config, ts.sys, process.cwd())
		this.program = ts.createProgram(parsed.fileNames, parsed.options)
		this.typeChecker = this.program.getTypeChecker()
	}

	compileProject(_entryFilePaths: string[]): CompilationOutput {
		// Discovery Pass
		this.discoveryPass()

		// Analysis Pass
		const blueprints: Record<string, WorkflowBlueprint> = {}
		const registry: Record<string, { importPath: string; exportName: string }> = {}
		const diagnostics: import('./types').CompilationDiagnostic[] = []

		for (const [_filePath, fileAnalysis] of this.fileCache) {
			for (const [exportName, { type, node }] of fileAnalysis.exports) {
				if (type === 'flow') {
					const analyzer = new FlowAnalyzer(this, fileAnalysis.sourceFile, node, this.typeChecker)
					const result = analyzer.analyze()
					blueprints[exportName] = result.blueprint
					Object.assign(registry, result.registry)
					diagnostics.push(...result.diagnostics)
				}
			}
		}

		// Generation Pass
		const manifestSource = this.generateManifest(blueprints, registry)

		return { blueprints, registry, diagnostics, manifestSource }
	}

	private discoveryPass(): void {
		for (const sourceFile of this.program.getSourceFiles()) {
			if (sourceFile.isDeclarationFile) continue
			const filePath = sourceFile.fileName
			const exports = new Map<string, { type: 'flow' | 'step'; node: ts.FunctionDeclaration }>()

			ts.forEachChild(sourceFile, (node) => {
				if (ts.isExportDeclaration(node)) {
					// Handle export { ... }
					if (node.exportClause && ts.isNamedExports(node.exportClause)) {
						node.exportClause.elements.forEach((element) => {
							const symbol = this.typeChecker.getSymbolAtLocation(element.name)
							if (symbol) {
								let originalSymbol = symbol
								if (symbol.flags & ts.SymbolFlags.Alias) {
									originalSymbol = this.typeChecker.getAliasedSymbol(symbol)
								}
								if (originalSymbol?.valueDeclaration && ts.isFunctionDeclaration(originalSymbol.valueDeclaration)) {
									const decl = originalSymbol.valueDeclaration
									const jsDocTags = ts.getJSDocTags(decl)
									const hasFlowTag = jsDocTags.some((tag) => tag.tagName.text === 'flow')
									const hasStepTag = jsDocTags.some((tag) => tag.tagName.text === 'step')
									if (hasFlowTag) {
										exports.set(element.name.text, { type: 'flow', node: decl })
									} else if (hasStepTag) {
										exports.set(element.name.text, { type: 'step', node: decl })
									}
								}
							}
						})
					}
				} else if (
					ts.isFunctionDeclaration(node) &&
					node.name &&
					ts.getCombinedModifierFlags(node) & ts.ModifierFlags.Export &&
					node.modifiers?.some((mod) => mod.kind === ts.SyntaxKind.AsyncKeyword)
				) {
					const jsDocTags = ts.getJSDocTags(node)
					const hasFlowTag = jsDocTags.some((tag) => tag.tagName.text === 'flow')
					const hasStepTag = jsDocTags.some((tag) => tag.tagName.text === 'step')
					if (hasFlowTag) {
						exports.set(node.name.text, { type: 'flow', node })
					} else if (hasStepTag) {
						exports.set(node.name.text, { type: 'step', node })
					}
				}
			})

			this.fileCache.set(filePath, { filePath, sourceFile, exports })
		}
	}

	private generateManifest(
		blueprints: Record<string, import('flowcraft').WorkflowBlueprint>,
		registry: Record<string, { importPath: string; exportName: string }>,
	): string {
		const imports: string[] = []
		const registryEntries: string[] = []

		const manifestDir = path.dirname(path.resolve('./dist/flowcraft.manifest.ts'))
		for (const [uses, { importPath, exportName }] of Object.entries(registry)) {
			const relativePath = path.relative(manifestDir, importPath).replace(/\.ts$/, '')
			imports.push(`import { ${exportName} } from '${relativePath}'`)
			registryEntries.push(`  '${uses}': ${exportName}`)
		}

		const blueprintEntries = Object.entries(blueprints)
			.map(([id, blueprint]) => `  '${id}': ${JSON.stringify(blueprint, null, 2)}`)
			.join(',\n')

		return `// Generated by @flowcraft/compiler
${imports.join('\n')}

import type { NodeImplementation, WorkflowBlueprint } from 'flowcraft'

export const registry: Record<string, NodeImplementation> = {
${registryEntries.join(',\n')}
}

export const blueprints: Record<string, WorkflowBlueprint> = {
${blueprintEntries}
}
`
	}
}
