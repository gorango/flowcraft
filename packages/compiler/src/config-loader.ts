import fs from 'node:fs'
import path from 'node:path'
import { transformSync } from 'esbuild'
import type { FlowcraftConfig } from './types'

/**
 * Finds and loads the Flowcraft configuration file from the project root.
 * Supports flowcraft.config.ts and flowcraft.config.js.
 */
export async function loadConfig(root: string = process.cwd()): Promise<FlowcraftConfig> {
	const tsConfigPath = path.resolve(root, 'flowcraft.config.ts')
	const jsConfigPath = path.resolve(root, 'flowcraft.config.js')

	let configPath: string | undefined
	let isTs = false

	if (fs.existsSync(tsConfigPath)) {
		configPath = tsConfigPath
		isTs = true
	} else if (fs.existsSync(jsConfigPath)) {
		configPath = jsConfigPath
	}

	if (!configPath) {
		return {} // No config file found
	}

	try {
		if (isTs) {
			// Transpile TypeScript config in memory
			const tsContent = fs.readFileSync(configPath, 'utf-8')
			const { code } = transformSync(tsContent, {
				loader: 'ts',
				format: 'esm',
			})

			// Use a data URI to import the transpiled code without writing to disk
			const dataUri = `data:text/javascript;base64,${Buffer.from(code).toString('base64')}`
			const module = await import(dataUri)
			return module.default || {}
		} else {
			// Directly import JavaScript config
			const module = await import(configPath)
			return module.default || {}
		}
	} catch (e) {
		console.error(`Error loading Flowcraft config file: ${configPath}`)
		throw e
	}
}
