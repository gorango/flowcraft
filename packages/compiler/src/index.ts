import { Compiler } from './compiler'
import type { CompilationOutput } from './types'

export function compileProject(entryFilePaths: string[], tsConfigPath: string): CompilationOutput {
	const compiler = new Compiler(tsConfigPath)
	return compiler.compileProject(entryFilePaths)
}
