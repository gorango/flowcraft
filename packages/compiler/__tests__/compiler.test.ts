import { describe, it, expect } from 'vitest'
import { Compiler } from '../src/compiler'
import * as path from 'path'

describe('Compiler', () => {
	it('should compile a simple project', () => {
		const compiler = new Compiler('tsconfig.json')
		const result = compiler.compileProject(['src/index.ts'])
		expect(result.blueprints).toBeDefined()
		expect(result.registry).toBeDefined()
		expect(result.diagnostics).toBeDefined()
	})

	it('should compile fixture with subflow', () => {
		const compiler = new Compiler('tsconfig.json')
		const result = compiler.compileProject(['__tests__/fixtures/main-flow.ts'])
		expect(result.blueprints).toMatchSnapshot()
		expect(result.registry).toMatchSnapshot()
		expect(result.manifestSource).toMatchSnapshot()
	})
})
