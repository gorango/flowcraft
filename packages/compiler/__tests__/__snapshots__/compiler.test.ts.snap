// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Compiler > should compile all test fixtures 1`] = `
{
  "mainFlow": {
    "edges": [
      {
        "source": "fetchUser_1",
        "target": "subFlow_1",
      },
      {
        "source": "subFlow_1",
        "target": "processOrders_1",
      },
    ],
    "id": "mainFlow",
    "nodes": [
      {
        "id": "fetchUser_1",
        "uses": "fetchUser",
      },
      {
        "id": "subFlow_1",
        "params": {
          "blueprintId": "subFlow",
        },
        "uses": "subflow",
      },
      {
        "id": "processOrders_1",
        "uses": "processOrders",
      },
    ],
  },
  "simpleFlow": {
    "edges": [
      {
        "source": "fetchUser_1",
        "target": "processOrders_1",
      },
    ],
    "id": "simpleFlow",
    "nodes": [
      {
        "id": "fetchUser_1",
        "uses": "fetchUser",
      },
      {
        "id": "processOrders_1",
        "uses": "processOrders",
      },
    ],
  },
  "subFlow": {
    "edges": [],
    "id": "subFlow",
    "nodes": [
      {
        "id": "recordTransaction_1",
        "uses": "recordTransaction",
      },
    ],
  },
}
`;

exports[`Compiler > should compile all test fixtures 2`] = `
{
  "fetchUser": {
    "exportName": "fetchUser",
    "importPath": "/home/g/m/flowcraft/packages/compiler/__tests__/fixtures/simple-flow.ts",
  },
  "processOrders": {
    "exportName": "processOrders",
    "importPath": "/home/g/m/flowcraft/packages/compiler/__tests__/fixtures/simple-flow.ts",
  },
  "recordTransaction": {
    "exportName": "recordTransaction",
    "importPath": "/home/g/m/flowcraft/packages/compiler/__tests__/fixtures/steps.ts",
  },
}
`;

exports[`Compiler > should compile all test fixtures 3`] = `
"// Generated by @flowcraft/compiler
import { recordTransaction } from '../__tests__/fixtures/steps'
import { fetchUser } from '../__tests__/fixtures/simple-flow'
import { processOrders } from '../__tests__/fixtures/simple-flow'

import type { NodeImplementation, WorkflowBlueprint } from 'flowcraft'

export const registry: Record<string, NodeImplementation> = {
  'recordTransaction': recordTransaction,
  'fetchUser': fetchUser,
  'processOrders': processOrders
}

export const blueprints: Record<string, WorkflowBlueprint> = {
  'subFlow': {
  "id": "subFlow",
  "nodes": [
    {
      "id": "recordTransaction_1",
      "uses": "recordTransaction"
    }
  ],
  "edges": []
},
  'mainFlow': {
  "id": "mainFlow",
  "nodes": [
    {
      "id": "fetchUser_1",
      "uses": "fetchUser"
    },
    {
      "id": "subFlow_1",
      "uses": "subflow",
      "params": {
        "blueprintId": "subFlow"
      }
    },
    {
      "id": "processOrders_1",
      "uses": "processOrders"
    }
  ],
  "edges": [
    {
      "source": "fetchUser_1",
      "target": "subFlow_1"
    },
    {
      "source": "subFlow_1",
      "target": "processOrders_1"
    }
  ]
},
  'simpleFlow': {
  "id": "simpleFlow",
  "nodes": [
    {
      "id": "fetchUser_1",
      "uses": "fetchUser"
    },
    {
      "id": "processOrders_1",
      "uses": "processOrders"
    }
  ],
  "edges": [
    {
      "source": "fetchUser_1",
      "target": "processOrders_1"
    }
  ]
}
}
"
`;

exports[`Compiler > should compile fixture with subflow 1`] = `
{
  "mainFlow": {
    "edges": [],
    "id": "mainFlow",
    "nodes": [
      {
        "id": "start",
        "uses": "start",
      },
    ],
  },
  "simpleFlow": {
    "edges": [
      {
        "source": "fetchUser_1",
        "target": "processOrders_1",
      },
    ],
    "id": "simpleFlow",
    "nodes": [
      {
        "id": "fetchUser_1",
        "uses": "fetchUser",
      },
      {
        "id": "processOrders_1",
        "uses": "processOrders",
      },
    ],
  },
  "subFlow": {
    "edges": [],
    "id": "subFlow",
    "nodes": [
      {
        "id": "start",
        "uses": "start",
      },
    ],
  },
}
`;

exports[`Compiler > should compile fixture with subflow 2`] = `
{
  "fetchUser": {
    "exportName": "fetchUser",
    "importPath": "/home/g/m/flowcraft/packages/compiler/__tests__/fixtures/simple-flow.ts",
  },
  "processOrders": {
    "exportName": "processOrders",
    "importPath": "/home/g/m/flowcraft/packages/compiler/__tests__/fixtures/simple-flow.ts",
  },
}
`;

exports[`Compiler > should compile fixture with subflow 3`] = `
"// Generated by @flowcraft/compiler
import { fetchUser } from '__tests__/fixtures/simple-flow.ts'
import { processOrders } from '__tests__/fixtures/simple-flow.ts'

import type { NodeImplementation, WorkflowBlueprint } from 'flowcraft'

export const registry: Record<string, NodeImplementation> = {
  'fetchUser': fetchUser,
  'processOrders': processOrders
}

export const blueprints: Record<string, WorkflowBlueprint> = {
  'subFlow': {
  "id": "subFlow",
  "nodes": [
    {
      "id": "start",
      "uses": "start"
    }
  ],
  "edges": []
},
  'mainFlow': {
  "id": "mainFlow",
  "nodes": [
    {
      "id": "start",
      "uses": "start"
    }
  ],
  "edges": []
},
  'simpleFlow': {
  "id": "simpleFlow",
  "nodes": [
    {
      "id": "fetchUser_1",
      "uses": "fetchUser"
    },
    {
      "id": "processOrders_1",
      "uses": "processOrders"
    }
  ],
  "edges": [
    {
      "source": "fetchUser_1",
      "target": "processOrders_1"
    }
  ]
}
}
"
`;
